### PR TITLE
crl-release-20.2: db: prevent conflicting, concurrent L0->LBase compactions

### LINKED ISSUE #875: db: concurrent, overlapping L0→Lbase compactions permitted
I was testing a nightly metamorphic tests TeamCity configuration and it uncovered an invariant violation.

https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_Pebble_Metamorphic/2249146:id/stress.log

```
[16:06:59][Step 1/1]             ===== SEED =====
[16:06:59][Step 1/1]             1599250004201769425
[16:06:59][Step 1/1]             ===== ERR =====
[16:06:59][Step 1/1]             exit status 1
[16:06:59][Step 1/1]             ===== OUT =====
[16:06:59][Step 1/1]             pebble: internal error: L5 files 000907 and 000908 have overlapping ranges: [jrcmlrtjrp#536,MERGE-jrcmlrtjrp#536,MERGE] vs [ijwnptq#535,MERGE-zxngb#72057594037927935,RANGEDEL]
[16:06:59][Step 1/1]             
[16:06:59][Step 1/1]             ===== OPTIONS =====
[16:06:59][Step 1/1]             [Version]
[16:06:59][Step 1/1]               pebble_version=0.1
[16:06:59][Step 1/1]             
[16:06:59][Step 1/1]             [Options]
[16:06:59][Step 1/1]               bytes_per_sync=32768
[16:06:59][Step 1/1]               cache_size=16777216
[16:06:59][Step 1/1]               cleaner=delete
[16:06:59][Step 1/1]               comparer=leveldb.BytewiseComparator
[16:06:59][Step 1/1]               delete_range_flush_delay=0s
[16:06:59][Step 1/1]               disable_wal=true
[16:06:59][Step 1/1]               flush_split_bytes=131072
[16:06:59][Step 1/1]               l0_compaction_concurrency=1
[16:06:59][Step 1/1]               l0_compaction_threshold=2
[16:06:59][Step 1/1]               l0_stop_writes_threshold=44
[16:06:59][Step 1/1]               l0_sublevel_compactions=true
[16:06:59][Step 1/1]               lbase_max_bytes=4096
[16:06:59][Step 1/1]               max_concurrent_compactions=3
[16:06:59][Step 1/1]               max_manifest_file_size=524288
[16:06:59][Step 1/1]               max_open_files=1000
[16:06:59][Step 1/1]               mem_table_size=1024
[16:06:59][Step 1/1]               mem_table_stop_writes_threshold=3
[16:06:59][Step 1/1]               min_compaction_rate=4194304
[16:06:59][Step 1/1]               min_flush_rate=1048576
[16:06:59][Step 1/1]               merger=pebble.concatenate
[16:06:59][Step 1/1]               table_property_collectors=[]
[16:06:59][Step 1/1]               wal_dir=wal
[16:06:59][Step 1/1]             
[16:06:59][Step 1/1]             [Level "0"]
[16:06:59][Step 1/1]               block_restart_interval=64
[16:06:59][Step 1/1]               block_size=512
[16:06:59][Step 1/1]               compression=Default
[16:06:59][Step 1/1]               filter_policy=none
[16:06:59][Step 1/1]               filter_type=table
[16:06:59][Step 1/1]               index_block_size=524288
[16:06:59][Step 1/1]               target_file_size=8388608
[16:06:59][Step 1/1]             
[16:06:59][Step 1/1]             [TestOptions]
[16:06:59][Step 1/1]               ingest_using_apply=true
```

### PR DESCRIPTION
**Pull request by jbowens**
Backport of #885.

----

Previously, it was possible for compaction picking to pick an L0->LBase
compaction that conflicted with an existing L0->LBase compaction under
specific conditions.

A flush of a new sstable constructs a new Version and sublevel
intervals. If a flushed table does not overlap any individual
currently-compacting files, its intervals within sublevels
datastructures are not marked as compacting, although the flushed table
may still overlap with the active compaction's key range (eg, if the new
sstable sits squarely between two compacting files' keyspaces). If no
files exist in LBase that intersect with the intersection of the newly
flushed sstable's keyspace and the active compaction's keyspace, a
new compaction of the flush sstable would be allowed to start.

This change augments our existing concurrent compaction failsafe to
compare picked L0->LBase compaction key ranges against the key ranges of
in-progress compactions, preventing the compaction if necessary.

Fix #875.

internal/manifest: mark full compaction range as base compacting

Update L0Sublevels to mark all the intervals contained within a base
compaction as base compacting, even if none of the intervals' files are
compacting. This is possible if a file was flushed, splitting existing
compacting intervals.

### CODE REVIEW

**Review by petermattis — APPROVED** [review]
:lgtm: (assuming this was a clean backport)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/935#-:-MHvu3CaAJfTLh-j3VYw:b1te5e8)* status: 0 of 9 files reviewed, all discussions resolved (waiting on @itsbilal and @petermattis)

**Review by jbowens — COMMENTED** [review]
> (assuming this was a clean backport)

Yeah, the SHA before the final rebase and merge of #885 cherry picked cleanly, so I used that.

TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/935#-:-MHvyd6WCsdEmvS-3cKe:b-dofhnr)* status: 0 of 9 files reviewed, all discussions resolved (waiting on @itsbilal)

**Pull request merged**
